lighthouse-core/gather/gatherers/seo/tap-targets.js

```diff
@@ -0,0 +1,343 @@
+    } = getComputedStyle(element);
+    if ((overflowX === 'hidden' && overflowY === 'hidden') || element.children.length === 0) {
+      // own size is 0x0 and there's no visible child content
+      return [];
+    }
+  }
+
+  // Treating overflowing content in scroll containers as invisible could mean that
+  // most of a given page is deemed invisible. But:
+  // - tap targets audit doesn't consider different containers/layers
+  // - having most content in an explicit scroll container is rare
+  // - treating them as hidden only generates false passes, which is better than false failures
+  clientRects = filterClientRectsWithinAncestorsVisibleScrollArea(element, clientRects);
+
+  return clientRects;
+}
+
+/**
+ *
+ * @param {Element} element
+ * @param {LH.Artifacts.Rect[]} clientRects
+ * @returns {LH.Artifacts.Rect[]}
+ */
+/* istanbul ignore next */
+function filterClientRectsWithinAncestorsVisibleScrollArea(element, clientRects) {
+  const parent = element.parentElement;
+  if (!parent) {
+    return clientRects;
+  }
+  if (getComputedStyle(parent).overflowY !== 'visible') {
+    const parentBCR = parent.getBoundingClientRect();
+    clientRects = clientRects.filter(cr => rectContains(parentBCR, cr));
+  }
+  if (parent.parentElement && parent.parentElement.tagName !== 'BODY') {
+    return filterClientRectsWithinAncestorsVisibleScrollArea(
+      parent,
+      clientRects
+    );
+  }
+  return clientRects;
+}
+
+/**
+ * @param {Element} element
+ * @returns {LH.Artifacts.Rect[]}
+ */
+/* istanbul ignore next */
+function getClientRects(element) {
+  const clientRects = Array.from(
+    element.getClientRects()
+  ).map(clientRect => {
+    // Contents of DOMRect get lost when returned from Runtime.evaluate call,
+    // so we convert them to plain objects.
+    const {width, height, left, top, right, bottom} = clientRect;
+    return {width, height, left, top, right, bottom};
+  });
+
+  for (const child of element.children) {
+    clientRects.push(...getClientRects(child));
+  }
+
+  return clientRects;
+}
+
+/**
+ * @param {Element} element
+ * @returns {boolean}
+ */
+/* istanbul ignore next */
+function elementHasAncestorTapTarget(element) {
+  // @ts-ignore
+  const realMatchesFn = window.__ElementMatches || window.Element.prototype.matches;
+  if (!element.parentElement) {
+    return false;
+  }
+  if (realMatchesFn.call(element.parentElement, tapTargetsSelector)) {
+    return true;
+  }
+  return elementHasAncestorTapTarget(element.parentElement);
+}
+
+/**
+ * Check if element is in a block of text, such as paragraph with a bunch of links in it.
+ * Makes a reasonable guess, but for example gets it wrong if the element is surrounded by other
+ * HTML elements instead of direct text nodes.
+ * @param {Element} element
+ * @returns {boolean}
+ */
+/* istanbul ignore next */
+function elementIsInTextBlock(element) {
+  /**
+   * @param {Element} element
+   */
+  function hasTextNodeSiblingsFormingTextBlock(element) {
```

Comment:
`hasTextNodeSiblingsFormingTextBlock` doesn't rely on being closed over by `elementIsInTextBlock`, so move outside to a top-level function?

Reply:
Makes sense, I was trying to avoid having another function that needs to be imported in the evalAsync string, but this is cleaner 👍